File: preprocess.py
#!/usr/bin/env python3
"""
Preprocess Research → docs/blog/assets.

Execute:

    # normal run
    env PYTHONUTF8=1 PYTHONHASHSEED=0 PYTHONDONTWRITEBYTECODE=1 python3 -OO preprocess.py

    # delete unused assets after build
    env PYTHONUTF8=1 PYTHONHASHSEED=0 PYTHONDONTWRITEBYTECODE=1 python3 -OO preprocess.py --clean
"""

from __future__ import annotations

import argparse
import concurrent.futures as cf
import json
import os
import re
import shutil
import unicodedata
import urllib.parse
from datetime import timezone
from pathlib import Path
from secrets import randbelow
from typing import (
    Any,
    Dict,
    Iterable,
    Iterator,
    List,
    Mapping,
    MutableMapping,
    Optional,
    Tuple,
    Union,
)

# ── constants ────────────────────────────────────────────────────────────────────

TZ_UTC = timezone.utc  # type: ignore
REPO = Path(__file__).resolve().parent
REPLACE_RULES_PATH = REPO / "replace_rules.json"

# ── helpers ──────────────────────────────────────────────────────────────────────


class CaseInsensitiveDict(MutableMapping[str, Any]):
    """A tiny case-insensitive dict."""

    def __init__(
        self,
        data: Optional[Union[Mapping[str, Any], Iterable[Tuple[str, Any]]]] = None,
        **kwargs: Any,
    ) -> None:
        self._store: Dict[str, Tuple[str, Any]] = {}
        data = data or {}
        self.update(data, **kwargs)

    def __getitem__(self, key: str) -> Any:
        return self._store[key.lower()][1]

    def __setitem__(self, key: str, value: Any) -> None:
        self._store[key.lower()] = (key, value)

    def __delitem__(self, key: str) -> None:
        del self._store[key.lower()]

    def __iter__(self) -> Iterator[str]:
        return (orig for orig, _ in self._store.values())

    def __len__(self) -> int:
        return len(self._store)


def nfc(text: str) -> str:
    """NFC-normalise a string."""
    return unicodedata.normalize("NFC", text)


def random_hex() -> str:
    """Return a cryptographically-safe 6-digit uppercase hex."""
    return f"{randbelow(0xFFFFFF + 1):06X}"


# ── load replace rules ───────────────────────────────────────────────────────────

with REPLACE_RULES_PATH.open("r", encoding="utf-8") as fh:
    REPLACE_RULES: Dict[str, str] = json.load(fh)

_REPLACE_RE = re.compile("|".join(map(re.escape, REPLACE_RULES)))
_LANG_FIX_RE = re.compile(r"---\s*\n(.*?)\n---", re.DOTALL)

# ── global regexes ───────────────────────────────────────────────────────────────

_WIKILINK_RE = re.compile(r"\[\[([^\]]+?)]]")              # raw [[…]] tokens
_CODE_BLOCK_RE = re.compile(r"```.*?```", re.S)            # fenced code
_IMG_RE = re.compile(r"!\[([^\]]*?)\]\(([^)]+?)\)$", re.M) # images
_SLUG_RE = re.compile(r"^slug:\s+['\"]?([^\s'\"#]+)['\"]?", re.M)

# ── markdown sanitisation ────────────────────────────────────────────────────────


def sanitise_md(root: Path) -> None:
    md_files = [p for p in root.rglob("*") if p.suffix in (".md", ".mdx")]
    print(f"📝 Sanitizing {len(md_files)} markdown files...")
    with cf.ThreadPoolExecutor() as ex:
        ex.map(_sanitise_one, md_files)
    print(f"✨ Completed markdown sanitization")


def _sanitise_one(path: Path) -> None:
    text = nfc(path.read_text("utf-8"))

    if "{{hex}}" in text and "template" not in path.name:
        text = text.replace("{{hex}}", "/" + random_hex())

    text = _REPLACE_RE.sub(lambda m: REPLACE_RULES[m.group(0)], text)

    fm = _LANG_FIX_RE.match(text)
    if fm and "lang: 'en'" in fm.group(1):
        if "div lang='ko" not in text and 'div lang="ko' not in text:
            ko_name = any("\uAC00" <= ch <= "\uD7A3" for ch in path.name)
            ko_chars = sum("\uAC00" <= ch <= "\uD7A3" for ch in text)
            en_chars = sum("a" <= ch.lower() <= "z" for ch in text)
            if ko_name or ko_chars > en_chars:
                text = text.replace("lang: 'en'", "lang: 'ko'")

    path.write_text(text, "utf-8")


# ── blog generation ──────────────────────────────────────────────────────────────


def process_blog(src: Path, en: Path, ko: Path, cfg: Path) -> None:
    print(f"📚 Processing blog content...")
    for dst in (en, ko):
        if dst.exists():
            shutil.rmtree(dst)
        dst.mkdir(parents=True)

    file_count = sum(1 for _ in src.rglob("*") if _.is_file())
    print(f"📋 Copying {file_count} blog files to English and Korean destinations")

    for entry in src.iterdir():
        if entry.is_file():
            shutil.copy(entry, en)
            shutil.copy(entry, ko)
        else:
            shutil.copytree(entry, en / entry.name)
            shutil.copytree(entry, ko / entry.name)

    shutil.copy(cfg / "english.yml", en / "authors.yml")
    shutil.copy(cfg / "korean.yml", ko / "authors.yml")
    _walk_rename(en, "en", "ko")
    _walk_rename(ko, "ko", "en")
    print(f"🌐 Completed blog processing")


def _walk_rename(base: Path, to_index: str, to_delete: str) -> None:
    for p in base.rglob("*"):
        if p.suffix not in (".md", ".mdx"):
            continue
        if p.stem.startswith(to_index):
            p.rename(p.with_name("index" + p.suffix))
        elif p.stem.startswith(to_delete):
            p.unlink()


# ── docs build ───────────────────────────────────────────────────────────────────


def process_docs(src: Path, dst: Path) -> None:
    print(f"📔 Processing documentation...")
    if dst.exists():
        shutil.rmtree(dst)
    shutil.copytree(src, dst)

    md_files = [p for p in dst.rglob("*.md")]
    print(f"🔗 Resolving links in {len(md_files)} markdown files")
    link_map = {nfc(p.stem.lower()): p for p in md_files}

    for yml in src.rglob("*.yml"):
        target = dst / yml.relative_to(src)
        target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy(yml, target)

    shutil.copytree(src / "assets", dst / "assets", dirs_exist_ok=True)

    with cf.ThreadPoolExecutor() as ex:
        ex.map(_process_images, md_files)
        ex.map(lambda p: _resolve_file(p, link_map), md_files)
    print(f"📘 Completed documentation processing")


def _process_images(path: Path) -> None:
    txt = path.read_text("utf-8")
    txt = re.sub(
        r"!\[\[([^]]+?)]]",
        lambda m: f"![{m.group(1)}](../assets/{m.group(1)})",
        txt,
    )
    path.write_text(txt, "utf-8")


def _resolve_file(path: Path, link_map: Dict[str, Path]) -> None:
    txt = path.read_text("utf-8")

    parts: List[str] = []
    last = 0
    for m in _CODE_BLOCK_RE.finditer(txt):
        start, end = m.span()
        outside = txt[last:start]
        outside = _WIKILINK_RE.sub(
            lambda w: _resolve_wikilink(w, path, link_map), outside
        )
        parts.append(outside)
        parts.append(m.group(0))
        last = end
    remainder = txt[last:]
    remainder = _WIKILINK_RE.sub(
        lambda w: _resolve_wikilink(w, path, link_map), remainder
    )
    parts.append(remainder)

    out = "".join(parts)
    if out != txt:
        path.write_text(out, "utf-8")


def _resolve_wikilink(match: re.Match[str], cur: Path, link_map: Dict[str, Path]) -> str:
    raw = match.group(1)

    # Skip tokens that are clearly not wiki titles
    if not raw or raw[0].isspace() or raw[-1].isspace() or raw.lstrip().startswith("-"):
        return match.group(0)  # leave untouched

    target, display = (raw.split("|", 1) + [raw])[:2]
    md = link_map.get(nfc(target.lower()))
    if not md:
        return match.group(0)  # unresolved → keep original

    rel = os.path.relpath(md, cur.parent)
    rel = "/".join(urllib.parse.quote(seg) for seg in rel.split("/"))
    return f"[{display}](./{rel})"


# ── backlink map ─────────────────────────────────────────────────────────────────


def build_backlinks(root: Path, out_dir: Path) -> None:
    print(f"🔄 Building backlink map...")
    backlink_map: Dict[str, Dict[str, str]] = CaseInsensitiveDict()
    uid_map: Dict[str, str] = CaseInsensitiveDict()

    file_count = 0
    link_count = 0
    
    for p in root.rglob("*.md"):
        file_count += 1
        fname = p.stem
        backlink_map.setdefault(fname, {})
        txt = p.read_text("utf-8")

        uid = _SLUG_RE.search(txt)
        if uid:
            uid_map[fname] = uid.group(1)

        for wl in _WIKILINK_RE.findall(txt):
            link_count += 1
            source = wl.split("|")[0]
            backlink_map.setdefault(source, {})[fname] = _context(txt, wl)

    out_dir.mkdir(parents=True, exist_ok=True)
    (out_dir / "backlinks.ts").write_text(
        "export const backlinks = "
        + json.dumps(dict(backlink_map), ensure_ascii=False, indent=4)
    )
    (out_dir / "filenames.ts").write_text(
        "export const filenames = "
        + json.dumps(dict(uid_map), ensure_ascii=False, indent=4)
    )
    print(f"🧩 Created backlink map with {file_count} files and {link_count} links")


def _context(txt: str, needle: str, keep: int = 6) -> str:
    tag = f"[[{needle}]]"
    for line in txt.splitlines():
        if tag not in line:
            continue
        pre_raw, post_raw = line.split(tag, 1)
        pre = " ".join(pre_raw.split()[-keep:])
        post = " ".join(post_raw.split()[:keep])
        return (
            ("... " + pre if pre_raw != pre else pre)
            + tag
            + (" " + post + "..." if post_raw != post else post)
        )
    return ""


# ── image alt fix ────────────────────────────────────────────────────────────────


def fix_img_alt(root: Path) -> None:
    count = 0
    for p in root.rglob("*"):
        if p.suffix not in (".md", ".mdx") or p.name == "Welcome.md":
            continue
        txt = p.read_text("utf-8")

        def repl(m: re.Match[str]) -> str:
            nonlocal count
            alt, src = m.groups()
            ext = Path(src).suffix.lower()
            count += 1
            if alt.endswith(ext) or alt.upper().startswith("ALT:"):
                clean = alt.replace("ALT:", "").strip()
                return f"\n<figure>\n\n![{clean}]({src})\n\n</figure>\n"
            return (
                f"\n<figure>\n\n![{alt}]({src})\n\n<figcaption>{alt}</figcaption>\n</figure>\n"
            )

        out = _IMG_RE.sub(repl, txt)
        if out != txt:
            p.write_text(out, "utf-8")


# ── asset cleanup ────────────────────────────────────────────────────────────────


def cleanup_assets(assets_dir: Path, research_root: Path) -> None:
    print(f"🧹 Checking for unused assets...")
    if not assets_dir.is_dir():
        return

    assets = {f.name for f in assets_dir.iterdir() if f.is_file()}
    print(f"🔍 Analyzing {len(assets)} assets for usage")
    mentioned = {f: False for f in assets}

    for md in research_root.rglob("*.md"):
        txt = md.read_text("utf-8")
        for a in assets:
            if a in txt:
                mentioned[a] = True

    unused = [a for a, used in mentioned.items() if not used]
    if not unused:
        print("✅ No unused assets found.")
        return

    print(f"🗑️ Found {len(unused)} unused assets:")
    for f in unused:
        print(" •", f)

    if input("Delete them? (y/N): ").lower().startswith("y"):
        for f in unused:
            (assets_dir / f).unlink()
            print(f"🗑️ Deleted {f}")


# ── entry point ─────────────────────────────────────────────────────────────────


def main() -> None:
    print("🚀 Starting preprocessing...")
    ap = argparse.ArgumentParser(description="Build site from Research markdown")
    ap.add_argument("--clean", action="store_true", help="Also delete unused assets")
    args = ap.parse_args()

    research = REPO / "Research"
    docs = REPO / "docs"
    blog_en = REPO / "blog"
    blog_ko = REPO / "i18n" / "ko" / "docusaurus-plugin-content-blog"
    posts_src = REPO / "posts"
    cfg = REPO / "config"
    out_ts = REPO / "src" / "data"
    assets = research / "assets"

    sanitise_md(research)
    process_blog(posts_src, blog_en, blog_ko, cfg)
    process_docs(research, docs)
    build_backlinks(docs, out_ts)
    fix_img_alt(docs)
    if args.clean:
        cleanup_assets(assets, research)

    print("✅ Preprocess completed.")


if __name__ == "__main__":
    main()

File: preprocess.ts
import { randomBytes } from 'node:crypto'
import { existsSync } from 'node:fs'
import {
  copyFile,
  mkdir,
  readdir,
  readFile,
  rename,
  rm,
  stat,
  unlink,
  writeFile,
} from 'node:fs/promises'
import {
  basename,
  dirname,
  extname,
  join,
  relative,
  resolve,
  sep,
} from 'node:path'
import { fileURLToPath } from 'node:url'
import { parseArgs } from 'node:util'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const REPO = resolve(__dirname)
const REPLACE_RULES_PATH = join(REPO, 'replace_rules.json')

class CaseInsensitiveMap<T> extends Map<string, T> {
  override get(key: string): T | undefined {
    return super.get(key.toLowerCase())
  }

  override set(key: string, value: T): this {
    return super.set(key.toLowerCase(), value)
  }

  override has(key: string): boolean {
    return super.has(key.toLowerCase())
  }

  override delete(key: string): boolean {
    return super.delete(key.toLowerCase())
  }
}

function nfc(text: string): string {
  return text.normalize('NFC')
}

function randomHex(): string {
  return randomBytes(3).toString('hex').toUpperCase()
}

const REPLACE_RULES: Record<string, string> = JSON.parse(
  await readFile(REPLACE_RULES_PATH, 'utf-8')
)

const REPLACE_RE = new RegExp(
  Object.keys(REPLACE_RULES)
    .map((key) => key.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
    .join('|'),
  'g'
)

const LANG_FIX_RE = /---\s*\n(.*?)\n---/s

const WIKILINK_RE = /\[\[([^\]]+?)]]/g
const CODE_BLOCK_RE = /```.*?```/gs
const IMG_RE = /!\[([^\]]*?)\]\(([^)]+?)\)$/gm
const SLUG_RE = /^slug:\s+['"]?([^\s'"#]+)['"]?/m
const WHITESPACE_RE = /\s+/

async function sanitiseMd(root: string): Promise<void> {
  const mdFiles = await findFiles(root, ['.md', '.mdx'])
  console.log(`📝 Sanitizing ${mdFiles.length} markdown files...`)

  await Promise.all(mdFiles.map(sanitiseOne))
  console.log('✨ Completed markdown sanitization')
}

async function sanitiseOne(filePath: string): Promise<void> {
  let text = nfc(await readFile(filePath, 'utf-8'))

  if (text.includes('{{hex}}') && !filePath.includes('template')) {
    text = text.replace('{{hex}}', `/${randomHex()}`)
  }

  text = text.replace(REPLACE_RE, (match) => REPLACE_RULES[match] ?? match)

  const fm = text.match(LANG_FIX_RE)
  if (
    fm?.[1]?.includes("lang: 'en'") &&
    !(text.includes("div lang='ko") || text.includes('div lang="ko'))
  ) {
    const fileName = basename(filePath)
    const koName = [...fileName].some((ch) => ch >= '\uAC00' && ch <= '\uD7A3')
    const koChars = [...text].filter(
      (ch) => ch >= '\uAC00' && ch <= '\uD7A3'
    ).length
    const enChars = [...text].filter(
      (ch) => (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')
    ).length

    if (koName || koChars > enChars) {
      text = text.replace("lang: 'en'", "lang: 'ko'")
    }
  }

  await writeFile(filePath, text, 'utf-8')
}

async function findFiles(dir: string, extensions: string[]): Promise<string[]> {
  const files: string[] = []

  async function walk(currentDir: string) {
    const entries = await readdir(currentDir, { withFileTypes: true })

    for (const entry of entries) {
      const fullPath = join(currentDir, entry.name)
      if (entry.isDirectory()) {
        await walk(fullPath)
      } else if (extensions.some((ext) => entry.name.endsWith(ext))) {
        files.push(fullPath)
      }
    }
  }

  await walk(dir)
  return files
}

async function copyRecursive(src: string, dest: string): Promise<void> {
  const srcStat = await stat(src)

  if (srcStat.isDirectory()) {
    await mkdir(dest, { recursive: true })
    const entries = await readdir(src)

    for (const entry of entries) {
      await copyRecursive(join(src, entry), join(dest, entry))
    }
  } else {
    await copyFile(src, dest)
  }
}

async function rmrf(dir: string): Promise<void> {
  if (existsSync(dir)) {
    await rm(dir, { recursive: true, force: true })
  }
}

async function processBlog(
  src: string,
  en: string,
  ko: string,
  cfg: string
): Promise<void> {
  console.log('📚 Processing blog content...')

  await rmrf(en)
  await rmrf(ko)
  await mkdir(en, { recursive: true })
  await mkdir(ko, { recursive: true })

  const entries = await readdir(src, { withFileTypes: true })
  const fileCount = await countFiles(src)
  console.log(
    `📋 Copying ${fileCount} blog files to English and Korean destinations`
  )

  for (const entry of entries) {
    const srcPath = join(src, entry.name)
    if (entry.isFile()) {
      await copyFile(srcPath, join(en, entry.name))
      await copyFile(srcPath, join(ko, entry.name))
    } else {
      await copyRecursive(srcPath, join(en, entry.name))
      await copyRecursive(srcPath, join(ko, entry.name))
    }
  }

  await copyFile(join(cfg, 'english.yml'), join(en, 'authors.yml'))
  await copyFile(join(cfg, 'korean.yml'), join(ko, 'authors.yml'))

  await walkRename(en, 'en', 'ko')
  await walkRename(ko, 'ko', 'en')

  console.log('🌐 Completed blog processing')
}

async function walkRename(
  base: string,
  toIndex: string,
  toDelete: string
): Promise<void> {
  const files = await findFiles(base, ['.md', '.mdx'])

  for (const file of files) {
    const dir = dirname(file)
    const fileBasename = basename(file, extname(file))
    const ext = extname(file)

    if (fileBasename.startsWith(toIndex)) {
      await rename(file, join(dir, `index${ext}`))
    } else if (fileBasename.startsWith(toDelete)) {
      await unlink(file)
    }
  }
}

async function countFiles(dir: string): Promise<number> {
  let count = 0
  const entries = await readdir(dir, { withFileTypes: true })

  for (const entry of entries) {
    const fullPath = join(dir, entry.name)
    if (entry.isDirectory()) {
      count += await countFiles(fullPath)
    } else {
      count++
    }
  }

  return count
}

async function processDocs(src: string, dst: string): Promise<void> {
  console.log('📔 Processing documentation...')

  await rmrf(dst)
  await copyRecursive(src, dst)

  const mdFiles = await findFiles(dst, ['.md'])
  console.log(`🔗 Resolving links in ${mdFiles.length} markdown files`)

  const linkMap = new CaseInsensitiveMap<string>()
  for (const file of mdFiles) {
    linkMap.set(nfc(basename(file, '.md')), file)
  }

  const ymlFiles = await findFiles(src, ['.yml'])
  for (const yml of ymlFiles) {
    const rel = relative(src, yml)
    const target = join(dst, rel)
    await mkdir(dirname(target), { recursive: true })
    await copyFile(yml, target)
  }

  const assetsDir = join(src, 'assets')
  if (existsSync(assetsDir)) {
    await copyRecursive(assetsDir, join(dst, 'assets'))
  }

  console.log(`📸 Processing images in ${mdFiles.length} files...`)
  await Promise.all(mdFiles.map((file) => processImages(file)))

  console.log('🔗 Resolving wikilinks...')
  await Promise.all(mdFiles.map((file) => resolveFile(file, linkMap)))

  console.log('📘 Completed documentation processing')
}

async function processImages(filePath: string): Promise<void> {
  let txt = await readFile(filePath, 'utf-8')

  txt = txt.replace(/!\[\[([^\]]+?)\]\]/g, (_match, p1) => {
    return `![${p1}](../assets/${p1})`
  })

  await writeFile(filePath, txt, 'utf-8')
}

async function resolveFile(
  filePath: string,
  linkMap: CaseInsensitiveMap<string>
): Promise<void> {
  const txt = await readFile(filePath, 'utf-8')
  const parts: string[] = []
  let lastIndex = 0

  const matches = [...txt.matchAll(CODE_BLOCK_RE)]

  for (const match of matches) {
    const start = match.index ?? 0
    const end = start + match[0].length

    const outside = txt.slice(lastIndex, start)
    const processedOutside = outside.replace(WIKILINK_RE, (m, p1) =>
      resolveWikilink(m, p1, filePath, linkMap)
    )

    parts.push(processedOutside)
    parts.push(match[0])
    lastIndex = end
  }

  const remainder = txt.slice(lastIndex)
  const processedRemainder = remainder.replace(WIKILINK_RE, (m, p1) =>
    resolveWikilink(m, p1, filePath, linkMap)
  )
  parts.push(processedRemainder)

  const out = parts.join('')
  if (out !== txt) {
    await writeFile(filePath, out, 'utf-8')
  }
}

function resolveWikilink(
  match: string,
  raw: string,
  currentFile: string,
  linkMap: CaseInsensitiveMap<string>
): string {
  if (
    !raw ||
    raw[0] === ' ' ||
    raw.at(-1) === ' ' ||
    raw.trimStart().startsWith('-')
  ) {
    return match
  }

  const pipeIndex = raw.indexOf('|')
  const target = pipeIndex === -1 ? raw : raw.slice(0, pipeIndex)
  const display = pipeIndex === -1 ? raw : raw.slice(pipeIndex + 1) || target
  const mdFile = linkMap.get(nfc(target))

  if (!mdFile) {
    return match
  }

  let rel = relative(dirname(currentFile), mdFile)
  rel = rel
    .split(sep)
    .map((segment) => encodeURIComponent(segment))
    .join('/')
  return `[${display}](./${rel})`
}

interface BacklinkMap {
  [key: string]: {
    [source: string]: string
  }
}

interface UidMap {
  [key: string]: string
}

function extractUid(txt: string): string | undefined {
  const uidMatch = txt.match(SLUG_RE)
  return uidMatch?.[1]
}

function processWikilinks(
  txt: string,
  fname: string,
  backlinkMap: BacklinkMap
): number {
  const wikilinks = [...txt.matchAll(WIKILINK_RE)]
  let linkCount = 0

  for (const match of wikilinks) {
    const rawWikilink = match[1]
    if (!rawWikilink) {
      continue
    }

    const pipeIndex = rawWikilink.indexOf('|')
    const source =
      pipeIndex === -1 ? rawWikilink : rawWikilink.slice(0, pipeIndex)
    if (!source) {
      continue
    }

    if (!backlinkMap[source]) {
      backlinkMap[source] = {}
    }

    backlinkMap[source][fname] = getContext(txt, rawWikilink)
    linkCount++
  }

  return linkCount
}

function sortBacklinkMap(backlinkMap: BacklinkMap): BacklinkMap {
  const sorted: BacklinkMap = {}

  for (const key of Object.keys(backlinkMap).sort()) {
    sorted[key] = {}
    const entries = backlinkMap[key]
    if (!entries) {
      continue
    }

    for (const nestedKey of Object.keys(entries).sort()) {
      const value = entries[nestedKey]
      if (value !== undefined) {
        sorted[key][nestedKey] = value
      }
    }
  }

  return sorted
}

function sortUidMap(uidMap: UidMap): UidMap {
  const sorted: UidMap = {}

  for (const key of Object.keys(uidMap).sort()) {
    const value = uidMap[key]
    if (value !== undefined) {
      sorted[key] = value
    }
  }

  return sorted
}

async function buildBacklinks(root: string, outDir: string): Promise<void> {
  console.log('🔄 Building backlink map...')

  const backlinkMap: BacklinkMap = {}
  const uidMap: UidMap = {}

  let fileCount = 0
  let linkCount = 0

  const allMdFiles = await findFiles(root, ['.md'])
  const mdFiles = allMdFiles.filter((file) => !file.includes('/templates/'))

  for (const file of mdFiles) {
    fileCount++
    const fname = basename(file, '.md')

    if (!backlinkMap[fname]) {
      backlinkMap[fname] = {}
    }

    const txt = await readFile(file, 'utf-8')

    const uid = extractUid(txt)
    if (uid) {
      uidMap[fname] = uid
    }

    linkCount += processWikilinks(txt, fname, backlinkMap)
  }

  await mkdir(outDir, { recursive: true })

  const sortedBacklinkMap = sortBacklinkMap(backlinkMap)
  const sortedUidMap = sortUidMap(uidMap)

  await writeFile(
    join(outDir, 'backlinks.json'),
    JSON.stringify(sortedBacklinkMap, null, 2)
  )

  await writeFile(
    join(outDir, 'filenames.json'),
    JSON.stringify(sortedUidMap, null, 2)
  )

  console.log(
    `🧩 Created backlink map with ${fileCount} files and ${linkCount} links`
  )
}

function getContext(txt: string, needle: string, keep = 6): string {
  const tag = `[[${needle}]]`
  const lines = txt.split('\n')

  for (const line of lines) {
    if (!line.includes(tag)) {
      continue
    }

    const [preRaw = '', postRaw = ''] = line.split(tag)
    const preParts = preRaw.split(WHITESPACE_RE)
    const postParts = postRaw.split(WHITESPACE_RE)

    const pre = preParts.slice(-keep).join(' ')
    const post = postParts.slice(0, keep).join(' ')

    return (
      (preRaw !== pre ? `... ${pre}` : pre) +
      tag +
      (postRaw !== post ? `${post} ...` : post)
    )
  }

  return ''
}

async function fixImgAlt(root: string): Promise<void> {
  const files = await findFiles(root, ['.md', '.mdx'])

  for (const file of files) {
    if (basename(file) === 'Welcome.md') {
      continue
    }

    const txt = await readFile(file, 'utf-8')

    const out = txt.replace(IMG_RE, (_match, alt, src) => {
      const ext = extname(src).toLowerCase()

      if (alt.endsWith(ext) || alt.toUpperCase().startsWith('ALT:')) {
        const clean = alt.replace('ALT:', '').trim()
        return `\n<figure>\n\n![${clean}](${src})\n\n</figure>\n`
      }

      return `\n<figure>\n\n![${alt}](${src})\n\n<figcaption>${alt}</figcaption>\n</figure>\n`
    })

    if (out !== txt) {
      await writeFile(file, out, 'utf-8')
    }
  }
}

async function cleanupAssets(
  assetsDir: string,
  researchRoot: string
): Promise<void> {
  console.log('🧹 Checking for unused assets...')

  if (!existsSync(assetsDir)) {
    return
  }

  const assets = await readdir(assetsDir)
  const assetFiles: string[] = []
  for (const asset of assets) {
    const assetStat = await stat(join(assetsDir, asset))
    if (assetStat.isFile()) {
      assetFiles.push(asset)
    }
  }
  console.log(`🔍 Analyzing ${assetFiles.length} assets for usage`)

  const mentioned = new Map<string, boolean>()
  for (const asset of assetFiles) {
    mentioned.set(asset, false)
  }

  const mdFiles = await findFiles(researchRoot, ['.md'])

  for (const mdFile of mdFiles) {
    const txt = await readFile(mdFile, 'utf-8')
    for (const asset of assetFiles) {
      if (txt.includes(asset)) {
        mentioned.set(asset, true)
      }
    }
  }

  const unused = assetFiles.filter((a) => !mentioned.get(a))

  if (unused.length === 0) {
    console.log('✅ No unused assets found.')
    return
  }

  console.log(`🗑️ Found ${unused.length} unused assets:`)
  for (const f of unused) {
    console.log(' •', f)
  }

  const readline = await import('node:readline')
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })

  const answer = await new Promise<string>((resolve) => {
    rl.question('Delete them? (y/N): ', resolve)
  })

  rl.close()

  if (answer.toLowerCase().startsWith('y')) {
    for (const f of unused) {
      await unlink(join(assetsDir, f))
      console.log(`🗑️ Deleted ${f}`)
    }
  }
}

async function main(): Promise<void> {
  console.log('🚀 Starting preprocessing...')

  const { values } = parseArgs({
    args: process.argv.slice(2),
    options: {
      clean: {
        type: 'boolean',
        default: false,
      },
    },
  })

  const research = join(REPO, 'Research')
  const docs = join(REPO, 'docs')
  const blogEn = join(REPO, 'blog')
  const blogKo = join(REPO, 'i18n', 'ko', 'docusaurus-plugin-content-blog')
  const postsSrc = join(REPO, 'posts')
  const cfg = join(REPO, 'config')
  const outTs = join(REPO, 'src', 'data')
  const assets = join(research, 'assets')

  await sanitiseMd(research)
  await processBlog(postsSrc, blogEn, blogKo, cfg)
  await buildBacklinks(research, outTs)
  await processDocs(research, docs)
  await fixImgAlt(docs)

  if (values.clean) {
    await cleanupAssets(assets, research)
  }

  console.log('✅ Preprocess completed.')
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error)
}
